### PR TITLE
V8 - Logging/Serilog specifics

### PR DESCRIPTION
**Pull request by warrenbuckley**
This PR contains detailed documentation about using Serilog in V8.
Currently this is a WIP/draft & will update this message or ask for feedback/review once I am done :)

### CODE REVIEW

**Comment by jmayntzhusen** [comment]
This already looks very great and informative! Can't wait to see when it is done! 😄

**Comment by warrenbuckley** [comment]
Thanks @jmayntzhusen 
I would like perhaps @marcemarc or @Jeavon to review this PR from a developer perspective & see if there would be any information that is missing or not clear.

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+```
+
+:::warning
+If you change the **serilog:minimum-level** to be **Error** then the following example above would only log messages from *Microsoft.AspNetCore.Mvc* and not any warning, info or debug messages from the *Microsoft* namespace
+:::
+
+### Changing the location of the main Umbraco Log file output
+*******WARREN TODO*******
+
+### Writing to a different source
```
Comment (Jeavon) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: I think writing to a source is a little confusing. How about "Writing log events to different storage types" or something along those lines?

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+If you change the main Umbraco logger in serilog.config to log only **Warning** you would not be able to have the serilog.user.config sub logger to be set to **Debug** as only Warning messages and higher will flow down into the child sub logger
+:::
+
+
+### Changing the log level for specific namespaces
+This can be done by adding the following into either `serilog.config` or the sub logger configuration file `serilog.user.config`
+
+```xml
+<add key="serilog:minimum-level:override:Microsoft" value="Warning" />
+<add key="serilog:minimum-level:override:Microsoft.AspNetCore.Mvc" value="Error" />
```
Comment (Jeavon) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: Might want to change this example to something frequently used with Umbraco (ImageProcessor maybe) as better to not get anyone excited at the mention of Core

**Comment by Jeavon** [comment]
@warrenbuckley overall this is looking awesome and I think it contains all the main areas I would be looking for. I've left a couple of minor line comments for your consideration! Looking forward to seeing the content expand further!

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 2 above)
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: Good point, it was just taken from Serilog docs when I implemented this. Happy for some suggestions @Jeavon

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+Logger.Info<MyApiController>($"We are saying hello to {name}");
+
+//Do not use :(
+Logger.Info(typeof(MyApiController), "We are saying hello to " + name);
+Logger.Info<MyApiController>("We are saying hello to " + name);
+```
+
+The above examples will write to the log file, however we will not get a seperate property logged with the message and we have no easy way to search for all log messages of this type.
+
+Where as the previous example we would be able to find all log messages that use the message template `We are saying hello to {Name}`
```
Comment (marcemarc) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: would it help when skimming this information to have the same pattern as the //Do Not Use ;-(, to reinforce the correct way?

eg

// Do Use :)
Logger.Info<MyApiController>("We are saying hello to {Name}", name);

it's like I immediately see the wrong way, but then can't find the right way, (although it is there!)

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+```xml
+<!-- Filters all sink's in the serilog.user.config to use this expression -->
+<!-- Common use case is to include SourceType starting with your own namespace -->
+<add key="serilog:using:FilterExpressions" value="Serilog.Filters.Expressions" />
+<add key="serilog:filter:ByIncluding.expression" value="StartsWith(SourceContext, 'MyNamespace')" />
+```
+
+### Writing log events to different storage types
+Serilog has a similar concept to Log4Net with its appenders, which are referred to Serilog Sinks.
```
Comment (jmayntzhusen) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]:
```suggestion
Serilog has a similar concept to Log4Net with its appenders, which are referred to as Serilog Sinks.
```

This is the name, right? 🙂

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+A Serilog Sink, allows you to persist the structured log message to a data store of your choice. In v8.0+ we use the *Serilog.Sinks.File* to allow us to write a .txt or .json file to disk. But the Serilog project anmd the wider Serilog community allows you to store these logs in various locations.
```
Comment (jmayntzhusen) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]:
```suggestion
A Serilog Sink, allows you to persist the structured log message to a data store of your choice. In v8.0+ we use the *Serilog.Sinks.File* to allow us to write a .txt or .json file to disk. But the Serilog project and the wider Serilog community allows you to store these logs in various locations.
```

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+For example you could install the Nuget package `PM> Install-Package Serilog.Sinks.Seq` and update the `serilog.user.config` file with the following XML snippet and if you already have the file example above it will write to that location as well as Seq.
+
+```xml
+<add key="serilog:using:Seq" value="Serilog.Sinks.Seq" />
+<add key="serilog:write-to:Seq.serverUrl" value="http://localhost:5341" />
+<add key="serilog:write-to:Seq.apiKey" value="[optional API key here]" />
+```
+
+### Adding a custom log property to all log items
+You may wish to add a log property to all log messages. A good example could be a Log property for the `environment` to determine if the log message came from `development` or `production`.
```
Comment (jmayntzhusen) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]:
```suggestion
You may wish to add a log property to all log messages. A good example could be a log property for the `environment` to determine if the log message came from `development` or `production`.
```

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+```
+
+The above examples will write to the log file, however we will not get a seperate property logged with the message and we have no easy way to search for all log messages of this type.
+
+Where as the previous example we would be able to find all log messages that use the message template `We are saying hello to {Name}`
+
+If you are writing custom code and wish to use the logger, it's available in the following places:
+* SurfaceController
+* UmbracoApiController
+* UmbracoAuthorizedApiController
```
Comment (marcemarc) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: not sure if this needs an example, it would be a repeat of the opening example I think, but maybe rewording this, just because it may not be clear 'how it's available' so more like...

If you are writing custom code and are inheriting from one of these special Umbraco base classes:

SurfaceController
UmbracoApiController
UmbracoAuthorizedApiController
RenderMvcController?

Then you can access the logging functionality via a special 'Logger' property included in those base classes (Do you need to add a using Umbraco.Core.Logging too?)

Outside of these places, eg a ContentFinder or your own custom code, you can get a reference to the logger from the Umbraco.Core.Composing namespace, using the 'Current.Logger' instance.

then I would put an example in of dong this...

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+1. **Warning** - indicators of possible issues or service/functionality degradation
+1. **Error** - indicating a failure within the application or connected system
+1. **Fatal** - critical errors causing complete failure of the application
+
+The default log levels we ship with in Umbraco v8.0+ are:
+* .txt file **Debug**
+* .json file **Verbose**
+
+## Configuration
+Serilog can be configured and extended by using the two XML configuration files on disk found at `/config/serilog.config` which is used to modify the main Umbraco logging piepleine and a second configuration file that is at `/config/serilog.user.config` which is a sublogger and allows you to make modifications without effecting the main Umbraco logger.
```
Comment (marcemarc) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: yumm piepleine

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+1. **Debug** - internal control flow and diagnostic state dumps to facilitate pinpointing of recognised problems
+1. **Information** - events of interest or that have relevance to outside observers; the default enabled minimum logging level
+1. **Warning** - indicators of possible issues or service/functionality degradation
+1. **Error** - indicating a failure within the application or connected system
+1. **Fatal** - critical errors causing complete failure of the application
+
+The default log levels we ship with in Umbraco v8.0+ are:
+* .txt file **Debug**
+* .json file **Verbose**
+
```
Comment (marcemarc) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: Do we want to put anything in here about over-logging? or is serilog super performant! Log4Net can cause havoc if it's set to Debug, and there is an abundance of debug messages logged...

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+A Serilog Sink, allows you to persist the structured log message to a data store of your choice. In v8.0+ we use the *Serilog.Sinks.File* to allow us to write a .txt or .json file to disk. But the Serilog project anmd the wider Serilog community allows you to store these logs in various locations.
+
+Here are some examples:
+* [Azure Table Storage](https://github.com/serilog/serilog-sinks-azuretablestorage)
+* [Elastic Search](https://github.com/serilog/serilog-sinks-elasticsearch)
+* [Elmah.io](https://github.com/serilog/serilog-sinks-elmahio)
+* [Seq](https://github.com/serilog/serilog-sinks-seq)
+* [SQL Server](https://github.com/serilog/serilog-sinks-mssqlserver)
+* [Windows Event Log](https://github.com/serilog/serilog-sinks-eventlog)
+* [Many more](https://github.com/serilog/serilog/wiki/Provided-Sinks)
```
Comment (dampee) [inline Getting-Started/Code/Debugging/Logging/index-v8.md]: I would add the least possible to this list.  They can look at all the provided sinks on the serilog page.  Otherwise this will become a magnet for link builders or "fans" adding their product.  I (e.g.) would add Exceptionless instead of Elmah.

Seq might be the only exception (as we are running Seq too...)

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 8 above)
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: Yum Yum @marcemarc

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 9 above)
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: I am not sure if or how much more performant Serilog is versus Log4Net and yet to have a site where I have heavily stressed tested this yet @marcemarc in the wild whilst V8 still in active dev.

What would happen with Log4Net when hammered with logs, does it cause the site to stop rendering or slow down requests or just kill your disk IO?!

We can write something about not over logging & I can change the defaults we ship in V8 to be higher if that is more generally prefered, but it would be good for me to test/verify somehow if this is more performant than Log4Net or not, but I would need to know what issues to look for/compare to.

File: Getting-Started/Code/Debugging/Logging/index-v8.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+versionFrom: 8.0.0
+---
+# Logging
+
+In Umbraco v8.0+ we have changed the underlying logging framework from Log4Net to Serilog.
+
+Out of the box for v8.0+ we will write two log files to disk, a .txt file in the same identical format as previously logged by Log4Net, so that you can continue to use any tools or regular expressions you may have used and secondly we will write a JSON file that contains a more rich logfile, that allows tools to perform searches & correlation on log patterns alot easier.
+
+The default location of these files are written to `App_Data/Logs` and contains the Machine name, along with the date too:
+* `/App_Data/Logs/UmbracoTraceLog.DELLBOOK.20181108.txt`
```
[the remark this answers is not in the document]
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: Yes I am not sure I can fully replicate the filename format that Log4Net does where it appends the date after the file extension like so

`UmbracoTraceLog.PrWeSWEs2V3AJ.txt.2018-10-10`

I could probably get the files in V8 to use hyphens if that is needed/required @jmayntzhusen ?!

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 2 above)
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: Done @Jeavon

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 10 above)
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: Good point, updated this @dampee

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jmayntzhusen) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: I think adding the hyphens improves readability a lot, so if it's not too bad then it would be great. Appending the date after the extension is super annoying anyways as you each time need to specify a programme to open that sort of files with, so if it could end up as `UmbracoTraceLog.DELLBOOK.2018-11-08.txt` then that would be perfect in my opinion! 😁

**Comment by warrenbuckley** [comment]
Thanks all for the great feedback & typo checking!
It's ready for a re-review please, fingers crossed your happy gang 
@jmayntzhusen @Jeavon @marcemarc @dampee 

Fingers Crossed

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 9 above)
Comment (marcemarc) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: @warrenbuckley yes, the site requests slow down, presuming because of disk IO, generally, now in production we run a site with logging levels set to ERROR, (small logs, see errors easily) -  then switch to DEBUG if there is an issue to debug, that way you can 'over log' in DEBUG, eg the more logging the better!, but without needlessly logging info when all is running fine. but it's a trade-off as I guess having that debug info there for people new to Umbraco from the get-go, means it's discoverable as a site is developed, ....so maybe it's less of a Tip for the documentation of serilog, and more relevant for a page on 'hardening umbraco for production' ?

Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: Yes perhaps it's a tip or health check instead?!
Some other Serilog sinks batch the logs and send them off to the destination, so another hardening tip could be to store logs in a different source than on disk.

What are your thoughts on the default log levels that are shipped in V8 for the txt & json file @marcemarc ?!

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: I agree its a bit more human readable but I can't see any easy way to configure or change this for the `Serilog.Sinks.File` package/project - https://github.com/serilog/serilog-sinks-file

So, for now, the date will have to be in this format of `20181010` @jmayntzhusen

File: Getting-Started/Code/Debugging/Logging/index-v8.md (hunk 9 above)
Comment (warrenbuckley) [reply Getting-Started/Code/Debugging/Logging/index-v8.md]: @jmayntzhusen I agree with you but the plan is to have an inbuilt logviewer built into the core of Umbraco so you can easily search & filter these log messages. It's currently a WIP and will be due for a tech review soon by the core D-Team

Here is an early sneak peek
image

image

**Comment by jmayntzhusen** [comment]
Since we have all reviewed at different times and Warren has made every single change we posted about I will merge this now 😃🎉

**Comment by warrenbuckley** [comment]
YAY - thanks all, if there is still anything not clear or needs better clarification, let me know and I will gladly fix it up :)
